spec/unit/agent_spec.rb

```diff
@@ -245,6 +245,22 @@ def controlled_run(&block)
           777
         }
       end
+
+      it "should return `1` exit code if the block returns `nil`" do
+        expect(Kernel).to receive(:fork).and_yield
+        expect(@agent).to receive(:exit).with(1)
```

Comment:
rspec has a builtin `exit_with` matcher for this:

```ruby
expect {
  @agent.run_in_fork {
    nil
  }
}.to exit_with(1)
```

Reply:
@joshcooper , if we use `exit_with` variant it will catch https://github.com/puppetlabs/puppet/blob/fe708fa5906effa0b3c60e3b5be75c7edcb34055/lib/puppet/agent.rb#L100 instead of https://github.com/puppetlabs/puppet/blob/fe708fa5906effa0b3c60e3b5be75c7edcb34055/lib/puppet/agent.rb#L98, it will always exit with -1, should I still keep all 3 tests then?

Reply:
Wow, this code doesn't work the way I thought it did... thanks for raising the question.

Originally puppet agent never forked, so the code looked like https://github.com/puppetlabs/puppet/commit/fa4a1d9e932db5e28c2654dadbf8402357271150#diff-ac44b8f38a4592c950df5cdc74689d93L45-L51. In the old code, we wanted to handle all exceptions except for `SystemExit` and `NoMemoryError`, so those were special cased. Makes sense.

But now that the agent forks, it doesn't make sense to rescue `SystemExit`/`NoMemoryError`. In all cases the forked child is going to exit. Worse, we always lose the child's exit code, since the child calls `exit(yield || 1)` but then rescues `SystemExit`, calls `exit(-1)`. Since the child process always exits with -1, then the parent ends up calling `raise SystemExit`, which [defaults the `status` field to `true`](https://ruby-doc.org/core-2.5.1/SystemExit.html

Shouldn't the `run_in_fork` method just do:

```ruby
atForkHandler = Puppet::Util::AtFork.get_handler
atForkHandler.prepare
begin
  child_pid = Kernel.fork do
    $0 = _("puppet agent: applying configuration")
    exit(yield || 1)
  end
ensure
  atForkHandler.parent
end

exit_code = Process.waitpid2(child_pid)
exit(exit_code[1].exitstatus)
```

I think this hasn't been noticed, because we don't record the exit code of a daemonized agent. It's only when running "onetime" that we return a "proper" exit code, and "onetime" sets `should_fork` to false, skipping the problem.

Reply:
I will apply this change and check if any side effect

Reply:
@joshcooper , I pushed the code with the changes, I'm not sure if we loose any useful functionality by removing `NoMemoryError` handling.

Reply:
The only loss that I can tell is previously if the child hit `NoMemoryError`, it exited with -2, which is at least on posix is converted to 254. So the parent's `case -2` statement is never reached, and the parent exits with 254. With this PR, the child will exit with 1, and the parent will exit with the same. To preserve the same behavior as before I think it'd be good to handle that in the child:

```ruby
  child_pid = Kernel.fork do
    begin
      $0 = _("puppet agent: applying configuration")
      exit(yield || 1)
    rescue NoMemoryError
      exit(-2) # or maybe make this explicit as "exit(254)"?
    end
  end
```